File: muraves/tracks_reconstruction/4_MainRecSof/ClusterLists.cc
#include <iostream>
#include <vector>
#include <numeric>
#include <algorithm>
#include "ClusterLists.h"
#include <random>
using namespace std;

///////// FUNCTION TO OBTAIN THE CLUSTER POSITION FROM A WEIGHTED MEAN ///////////
double ClusterPosition(vector <double> stripDeposits, vector <double> stripPos) {
  double clusterPositionNum = 0.;
  double clusterPositionDen = 0.;
  double clusterPosition, addFactor;
  for(int i=0; i<stripDeposits.size(); i++) {
    addFactor = stripDeposits.at(i)*stripPos.at(i);
    clusterPositionNum = clusterPositionNum + addFactor;
    clusterPositionDen = clusterPositionDen + stripDeposits.at(i);
  }
  clusterPosition = (double) (clusterPositionNum/clusterPositionDen);
  return clusterPosition;
}

//////// FUCTION TO CALCULATE THE ENERGY OF THE CLUSTER (SUM OF THE STRIP DEPOSITS) ///////
double ClusterEnergy(vector <double> stripDeposits) {
  double ClusterEnergy =0;
   for(int i=0; i<stripDeposits.size(); i++) {
     ClusterEnergy += stripDeposits.at(i);
       }
   return  ClusterEnergy;
}

vector<int> SortIndices(vector<double> ReferenceVector) {

  vector<int> indices(ReferenceVector.size());
  iota(indices.begin(), indices.end(), 0);
  sort(indices.begin(), indices.end(),
           [&](int A, int B) -> bool {
                return ReferenceVector[A] > ReferenceVector[B];
            });
  return indices;
}

///////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////

ClusterCollection CreateClusterList(vector <double> Deposits, const double EnergyThreshold_clusterStrip, const double EnergyThreshold_singleStrip, double AdStripsThEnergy_singleStripCl,double T_exp1,double T_exp2,vector<double>TriggerMask1, vector<double>TriggerMask2) {
  //EXTERNAL PARAMETERS  
  double MaxStripEnergy = 3000;
  double FirstStripPos = -0.528;
  double AdiacentStripsDistance = 0.0165;
  const int nStrips = 64;
  // INTERNAL OBJECTS 
  vector <double> clusterStripsDeposits;
  vector <double> clusterStripsPos;
  vector <double> clusterStripsID;
  double deposit,cluster_position,cluster_energy;
  double Texp; 
  /// OUTPUT OBJECTS
  vector <double > clusterPosition, clusterEnergy, clusterSize;
  vector <vector<double>> Clusters_StripsDeposits,Clusters_StripsPositions, Clusters_StripsID;
  //////////////////
  
  // SORTING OBJECTS ---> sort strips and clusters according to the energy (from highest to lowest) 
  vector <int> OrderedStripsIndices;
  vector <double> SortedStipsPositions;
  vector <double> SortedStipsID;
  vector <double> SortedStipsDeposits;
  vector <int> Sorted_clusterSize;
  vector <int> OrderedClustersIndices;
  vector <double> SortedClustersPositions;
  vector <double> SortedClustersEnergies;
  vector <vector<double>> Sorted_Clusters_StripsDeposits,Sorted_Clusters_StripsPositions, Sorted_Clusters_StripsID;
  vector<double> Texp_cluster;
  int minEnergyTriggerMask = 20;
  int prev_strip=-1, subs_strip;
  double strip_position, pre_strip_position, post_strip_position;
  double cluEn=0, cluPos = 0;
  int cluSize=0;
  vector <int> strip_index;
  /// LOOP OVER THE 64 SCINTILLATOR BARS ////
  bool ClusterIsOn=1; 
  for(int st=0; st<nStrips; st++) {
    deposit = Deposits.at(st);
    if(deposit < EnergyThreshold_clusterStrip)
      ClusterIsOn=0;
    else ClusterIsOn=1;
    if(deposit >MaxStripEnergy)
      ClusterIsOn=0;

    if(deposit > minEnergyTriggerMask) {
      if(st<32) {

	if(std::find(TriggerMask1.begin(),TriggerMask1.end(),st)==TriggerMask1.end()) ClusterIsOn=0 ;}
      
      else {

	if(std::find(TriggerMask2.begin(),TriggerMask2.end(),st-32)==TriggerMask2.end())
	  ClusterIsOn=0;
      }
    }
    if(st == nStrips-1)
      ClusterIsOn=0;
    //if( ((st<32 && std::find(TriggerMask1.begin(),TriggerMask1.end(),st)!=TriggerMask1.end()) || (st>31 && std::find(TriggerMask2.begin(),TriggerMask2.end(),st-32)!=TriggerMask2.end())) &&  deposit > EnergyThreshold_clusterStrip && deposit < MaxStripEnergy) {

    if(ClusterIsOn==1) { 
      strip_position = (double) (st*AdiacentStripsDistance);
      /// Strip Informations //////////////////
      clusterStripsDeposits.push_back(deposit);
      clusterStripsPos.push_back(FirstStripPos + strip_position);
      clusterStripsID.push_back(st);
      /////////////////////////////////////
      //// Cluster energy and position //// 
      cluEn=cluEn+deposit;
      cluPos = cluPos + (strip_position +FirstStripPos)*deposit;
      /////////////////////////////////////

      cluSize++;
      strip_index.push_back(st);
      
    }
    else {
      ///////////////// IF cluster size is equal to 1 --> add adjacent strips is energetic ////
	if(cluSize==1) {
	  if(cluEn>EnergyThreshold_singleStrip && cluEn < MaxStripEnergy)  {
	    if(strip_index.at(0)<nStrips-1 && Deposits.at(1+strip_index.at(0))>AdStripsThEnergy_singleStripCl && Deposits.at(1+strip_index.at(0)) < MaxStripEnergy) {
	      post_strip_position = (double) ((strip_index.at(0)+1)*AdiacentStripsDistance);
	      cluPos = cluPos+ Deposits.at(1+strip_index.at(0))*(FirstStripPos + post_strip_position);
	      cluEn+=Deposits.at(strip_index.at(0)+1);
	      cluSize++;
	      clusterStripsDeposits.push_back(Deposits.at(strip_index.at(0)+1));
	      clusterStripsPos.push_back(FirstStripPos + post_strip_position);
	      clusterStripsID.push_back(strip_index.at(0)+1);
	    }
	    
	    if(strip_index.at(0)>0 && Deposits.at(strip_index.at(0)-1)>AdStripsThEnergy_singleStripCl  && Deposits.at(strip_index.at(0)-1)< MaxStripEnergy) {
	      pre_strip_position = (double) ((strip_index.at(0)-1)*AdiacentStripsDistance);
	      cluPos = cluPos+ Deposits.at(strip_index.at(0)-1)*(FirstStripPos + pre_strip_position);
	      cluEn+=Deposits.at(strip_index.at(0)-1);
	      clusterStripsDeposits.push_back(Deposits.at(strip_index.at(0)-1));
	      clusterStripsPos.push_back(FirstStripPos + pre_strip_position);
	      clusterStripsID.push_back(strip_index.at(0)-1);
	      cluSize++;
	    }

	  }
	  else {
	    cluPos=0;
	    cluEn=0;
	    cluSize=0;
	  }

	}
	if(cluSize>0) {
	  cluster_position = (double) cluPos/cluEn;
	  /////// Generate a random position within [-Strip width, + Strip_width] for those clusters with one strip
	  if(cluSize==1) {
	    const int range_from  = - (int) 1000*AdiacentStripsDistance/2;
	    const int range_to    = (int) 1000*AdiacentStripsDistance/2;
	    random_device                  rand_dev;
	    mt19937                        generator(rand_dev());
	    uniform_int_distribution<int>  distr(range_from, range_to);	   
	    cluster_position = cluster_position + (double) (distr(generator))/1000.;
	  }
	  cluster_energy = cluEn;
	  clusterPosition.push_back(cluster_position);
	  clusterEnergy.push_back(cluster_energy);
	  clusterSize.push_back(cluSize);
	  
	  // ORDER STRIPS ACCORDING TO INCREASING DEPOSITS ////////////////////////////////
	  OrderedStripsIndices.clear();
	  OrderedStripsIndices = SortIndices(clusterStripsDeposits);
	  
	  for(int ind=0; ind<cluSize; ind++) {
	    SortedStipsID.push_back(clusterStripsID.at(OrderedStripsIndices.at(ind)));
	    SortedStipsPositions.push_back(clusterStripsPos.at(OrderedStripsIndices.at(ind)) );
	    SortedStipsDeposits.push_back(clusterStripsDeposits.at(OrderedStripsIndices.at(ind)) );
	    
	  }
	  Clusters_StripsDeposits.push_back(SortedStipsDeposits);
	  Clusters_StripsPositions.push_back(SortedStipsPositions);
	  Clusters_StripsID.push_back(SortedStipsID);
	  ////////////////////////////////////////////////////////////////////////////
	  
    ////////////////////////////////
	}
	strip_index.clear();
	clusterStripsDeposits.clear();
	clusterStripsPos.clear();
	SortedStipsDeposits.clear();
	SortedStipsPositions.clear();
	SortedStipsID.clear();
	cluEn=0;
	cluPos=0;
	cluSize=0;	
    }
    
  }
  // ORDER CLUSTERS ACCORDING TO INCREASING ENERGY
  OrderedClustersIndices.clear();
  OrderedClustersIndices = SortIndices(clusterEnergy);

  for(int ind=0; ind<clusterEnergy.size(); ind++) {

    SortedClustersPositions.push_back(clusterPosition.at(OrderedClustersIndices.at(ind)));
    SortedClustersEnergies.push_back(clusterEnergy.at(OrderedClustersIndices.at(ind)));
    Sorted_clusterSize.push_back(clusterSize.at(OrderedClustersIndices.at(ind)));
    Sorted_Clusters_StripsPositions.push_back(Clusters_StripsPositions.at(OrderedClustersIndices.at(ind)));
    Sorted_Clusters_StripsDeposits.push_back(Clusters_StripsDeposits.at(OrderedClustersIndices.at(ind)));
    Sorted_Clusters_StripsID.push_back(Clusters_StripsID.at(OrderedClustersIndices.at(ind)));
    if(clusterPosition.at(OrderedClustersIndices.at(ind)) <= FirstStripPos + 31*AdiacentStripsDistance) {
	Texp = T_exp1;
      }
      else {
	Texp = T_exp2;
      }
    if(clusterPosition.at(OrderedClustersIndices.at(ind)) > FirstStripPos + 31*AdiacentStripsDistance &&clusterPosition.at(OrderedClustersIndices.at(ind)) < FirstStripPos + 32*AdiacentStripsDistance ) {
      if(T_exp1>T_exp2) Texp = T_exp1;
      else  Texp = T_exp2;
    }
    Texp_cluster.push_back(Texp);
  }
  
  int Nclusters = clusterPosition.size();
  ClusterCollection results;
  results.ClustersEnergy = SortedClustersEnergies;
  results.ClustersPositions =  SortedClustersPositions;
  results.ClustersSize = Sorted_clusterSize;
  results.StripsEnergy = Sorted_Clusters_StripsDeposits;
  results.StripsPositions = Sorted_Clusters_StripsPositions;
  results.TimeExpansions = Texp_cluster;
  results.StripsID = Sorted_Clusters_StripsID;
  return results;
}
     	  


    

       
	

File: muraves/tracks_reconstruction/4_MainRecSof/ClusterLists.py
from __future__ import annotations

from dataclasses import dataclass, field
import random
from typing import Sequence

from reco_config import get_reco_config


@dataclass
class ClusterCollection:
    ClustersEnergy: list[float] = field(default_factory=list)
    ClustersPositions: list[float] = field(default_factory=list)
    StripsEnergy: list[list[float]] = field(default_factory=list)
    StripsPositions: list[list[float]] = field(default_factory=list)
    StripsID: list[list[float]] = field(default_factory=list)
    TimeExpansions: list[float] = field(default_factory=list)
    ClustersSize: list[int] = field(default_factory=list)


class DeterministicSmearingRNG:
    """Cross-language RNG for reproducible single-strip smearing."""

    def __init__(self, seed: int):
        state = int(seed) & 0xFFFFFFFF
        self._state = state if state != 0 else 0x6D2B79F5

    def _next_u32(self) -> int:
        state = self._state
        state ^= (state << 13) & 0xFFFFFFFF
        state ^= (state >> 17) & 0xFFFFFFFF
        state ^= (state << 5) & 0xFFFFFFFF
        self._state = state & 0xFFFFFFFF
        return self._state

    def randint(self, low: int, high: int) -> int:
        if high < low:
            raise ValueError("high must be >= low")
        return low + (self._next_u32() % (high - low + 1))


def ClusterPosition(stripDeposits: Sequence[float], stripPos: Sequence[float]) -> float:
    """Return weighted cluster position from strip deposits and positions."""
    numerator = 0.0
    denominator = 0.0
    for deposit, position in zip(stripDeposits, stripPos):
        numerator += deposit * position
        denominator += deposit
    return numerator / denominator


def ClusterEnergy(stripDeposits: Sequence[float]) -> float:
    """Return total cluster energy as sum of strip deposits."""
    return float(sum(stripDeposits))


def SortIndices(referenceVector: Sequence[float]) -> list[int]:
    """Sort indices by decreasing values of the reference vector."""
    return sorted(range(len(referenceVector)), key=lambda idx: referenceVector[idx], reverse=True)


def CreateClusterList(
    Deposits: Sequence[float],
    EnergyThreshold_clusterStrip: float,
    EnergyThreshold_singleStrip: float,
    AdStripsThEnergy_singleStripCl: float,
    Texp1: float,
    Texp2: float,
    TriggerMask1: Sequence[float],
    TriggerMask2: Sequence[float],
    smearing_rng: DeterministicSmearingRNG | None = None,
) -> ClusterCollection:
    """
    Python port of CreateClusterList from ClusterLists.cc.

    The implementation intentionally mirrors the C++ logic, including cluster
    ordering by energy and the single-strip random position smearing.
    """
    cfg = get_reco_config()["cluster_lists"]
    max_strip_energy = float(cfg["max_strip_energy"])
    first_strip_pos = float(cfg["first_strip_pos"])
    adjacent_strips_distance = float(cfg["adjacent_strips_distance"])
    n_strips = int(cfg["n_strips"])
    min_energy_trigger_mask = float(cfg["min_energy_trigger_mask"])
    split_board_index = int(cfg["split_board_index"])
    smearing_scale = int(cfg["single_strip_smearing_scale"])

    cluster_strips_deposits: list[float] = []
    cluster_strips_pos: list[float] = []
    cluster_strips_id: list[float] = []

    cluster_positions: list[float] = []
    cluster_energies: list[float] = []
    cluster_sizes: list[int] = []

    clusters_strips_deposits: list[list[float]] = []
    clusters_strips_positions: list[list[float]] = []
    clusters_strips_id: list[list[float]] = []

    strip_index: list[int] = []
    clu_en = 0.0
    clu_pos = 0.0
    clu_size = 0

    # Preserve the old behavior unless a deterministic cross-language RNG is provided.
    nondeterministic_rng = random.SystemRandom() if smearing_rng is None else None

    for st in range(n_strips):
        deposit = float(Deposits[st])

        if deposit < EnergyThreshold_clusterStrip:
            cluster_is_on = False
        else:
            cluster_is_on = True

        if deposit > max_strip_energy:
            cluster_is_on = False

        if deposit > min_energy_trigger_mask:
            # condition regarding strips of the first borad (0-31) 
            if st < split_board_index:
                # if the strips is not in the trigger mask of the first board, there is no cluster.
                if st not in TriggerMask1:
                    cluster_is_on = False
            # condition regarding strips belonging to the second board (32-63)
            else:
                # if the strips is not in the trigger mask of the second board, there is no cluster.
                if (st - split_board_index) not in TriggerMask2:
                    cluster_is_on = False
        
        # (?) In the C++ code, there is an additional condition that checks if the strip index is the last one (n_strips - 1) and if so, it turns off the cluster.
        if st == (n_strips - 1):
            cluster_is_on = False

        if cluster_is_on:
            strip_position = st * adjacent_strips_distance
            cluster_strips_deposits.append(deposit)
            cluster_strips_pos.append(first_strip_pos + strip_position)
            cluster_strips_id.append(float(st))
            clu_en += deposit
            # cluster position is calculated as a weighted average of the strip positions, where the weights are the strip deposits.
            clu_pos += (strip_position + first_strip_pos) * deposit 
            clu_size += 1
            strip_index.append(st)
        else:
            # processing the case of single-strip cluster, which is a special case in the C++ code.
            if clu_size == 1:
                if EnergyThreshold_singleStrip < clu_en < max_strip_energy:
                    base_strip = strip_index[0]
                    # Checking strips adjacent to the single-strip cluster to see if they can be included in the cluster based on their energy deposits.
                    # Strip adjacent to the right (base_strip + 1)
                    if (
                        base_strip < (n_strips - 1)
                        and Deposits[base_strip + 1] > AdStripsThEnergy_singleStripCl
                        and Deposits[base_strip + 1] < max_strip_energy
                    ):
                        post_strip_position = (base_strip + 1) * adjacent_strips_distance
                        clu_pos += Deposits[base_strip + 1] * (first_strip_pos + post_strip_position)
                        clu_en += Deposits[base_strip + 1]
                        clu_size += 1
                        cluster_strips_deposits.append(float(Deposits[base_strip + 1]))
                        cluster_strips_pos.append(first_strip_pos + post_strip_position)
                        cluster_strips_id.append(float(base_strip + 1))
                    # Strip adjacent to the left (base_strip - 1)
                    if (
                        base_strip > 0
                        and Deposits[base_strip - 1] > AdStripsThEnergy_singleStripCl
                        and Deposits[base_strip - 1] < max_strip_energy
                    ):
                        pre_strip_position = (base_strip - 1) * adjacent_strips_distance
                        clu_pos += Deposits[base_strip - 1] * (first_strip_pos + pre_strip_position)
                        clu_en += Deposits[base_strip - 1]
                        cluster_strips_deposits.append(float(Deposits[base_strip - 1]))
                        cluster_strips_pos.append(first_strip_pos + pre_strip_position)
                        cluster_strips_id.append(float(base_strip - 1))
                        clu_size += 1
                else:
                    clu_pos = 0.0
                    clu_en = 0.0
                    clu_size = 0

            if clu_size > 0:
                # The cluster position is calculated as the weighted average of the strip positions, where the weights are the strip deposits.
                # Here, we divide the accumulated position by the total energy to get the final cluster position to complete the weighted average calculation.
                cluster_position = clu_pos / clu_en
                if clu_size == 1:
                    # Keep C++ granularity: integer millimeter steps after 1/1000 scaling.
                    range_from = int(-smearing_scale * adjacent_strips_distance / 2)
                    range_to = int(smearing_scale * adjacent_strips_distance / 2)
                    rng = smearing_rng if smearing_rng is not None else nondeterministic_rng
                    cluster_position += rng.randint(range_from, range_to) / float(smearing_scale)

                cluster_positions.append(cluster_position)
                cluster_energies.append(clu_en)
                cluster_sizes.append(clu_size)

                ordered_strip_indices = SortIndices(cluster_strips_deposits)
                sorted_strip_ids = [cluster_strips_id[idx] for idx in ordered_strip_indices]
                sorted_strip_positions = [cluster_strips_pos[idx] for idx in ordered_strip_indices]
                sorted_strip_deposits = [cluster_strips_deposits[idx] for idx in ordered_strip_indices]

                clusters_strips_deposits.append(sorted_strip_deposits)
                clusters_strips_positions.append(sorted_strip_positions)
                clusters_strips_id.append(sorted_strip_ids)

            strip_index.clear()
            cluster_strips_deposits.clear()
            cluster_strips_pos.clear()
            cluster_strips_id.clear()
            clu_en = 0.0
            clu_pos = 0.0
            clu_size = 0

    ordered_cluster_indices = SortIndices(cluster_energies)

    sorted_cluster_positions: list[float] = []
    sorted_cluster_energies: list[float] = []
    sorted_cluster_size: list[int] = []
    sorted_clusters_strips_positions: list[list[float]] = []
    sorted_clusters_strips_deposits: list[list[float]] = []
    sorted_clusters_strips_id: list[list[float]] = []
    texp_cluster: list[float] = []

    split_boundary_low = first_strip_pos + (split_board_index - 1) * adjacent_strips_distance
    split_boundary_high = first_strip_pos + split_board_index * adjacent_strips_distance

    for idx in ordered_cluster_indices:
        pos = cluster_positions[idx]
        sorted_cluster_positions.append(pos)
        sorted_cluster_energies.append(cluster_energies[idx])
        sorted_cluster_size.append(cluster_sizes[idx])
        sorted_clusters_strips_positions.append(clusters_strips_positions[idx])
        sorted_clusters_strips_deposits.append(clusters_strips_deposits[idx])
        sorted_clusters_strips_id.append(clusters_strips_id[idx])

        if pos <= split_boundary_low:
            texp = Texp1
        else:
            texp = Texp2

        if split_boundary_low < pos < split_boundary_high:
            texp = Texp1 if Texp1 > Texp2 else Texp2

        texp_cluster.append(texp)

    return ClusterCollection(
        ClustersEnergy=sorted_cluster_energies,
        ClustersPositions=sorted_cluster_positions,
        StripsEnergy=sorted_clusters_strips_deposits,
        StripsPositions=sorted_clusters_strips_positions,
        StripsID=sorted_clusters_strips_id,
        TimeExpansions=texp_cluster,
        ClustersSize=sorted_cluster_size,
    )


# Pythonic aliases
cluster_position = ClusterPosition
cluster_energy = ClusterEnergy
sort_indices = SortIndices
create_cluster_list = CreateClusterList
